File: 0007_reverse_integer/reverse_integer.java
public class Solution {
    public int reverse(int x) {
        int res = 0;
        while (x != 0) {
            int remainder = x % 10;
            int tmp = res * 10 + remainder;
            // if overflow happens, then next statement will be true
            if ((tmp - remainder) / 10 != res) {
                return 0;
            }
            res = tmp;
            x /= 10;
        }
        return res;
    }
}

File: 0007_reverse_integer/reverse_integer.rb
def reverse(x)
    sign = x > 0 ? 1 : -1
    x = x.abs
    res = 0
    until x.zero?
        remainder = x.remainder(10)
        x /= 10
        res = (res * 10) + remainder
    end
    return 0 if res < int_min || res > int_max
    res * sign
end

def int_max
    2**31 - 1
end

def int_min
    -(2**31)
end